internal/mongoutil/mongoutil.go

```diff
@@ -0,0 +1,125 @@
+// a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+
+package mongoutil
+
+import (
+	"reflect"
+
+	"go.mongodb.org/mongo-driver/mongo/options"
+)
+
+// Args defines arguments types that can be merged using the functional setters.
+type Args interface {
+	options.AggregateArgs | options.BucketArgs | options.BulkWriteArgs |
+		options.ClientArgs | options.ClientEncryptionArgs | options.CollectionArgs |
+		options.CountArgs | options.CreateIndexesArgs |
+		options.CreateCollectionArgs | options.CreateSearchIndexesArgs |
+		options.CreateViewArgs | options.DataKeyArgs | options.DatabaseArgs |
+		options.DefaultIndexArgs | options.DeleteArgs | options.DistinctArgs |
+		options.DropCollectionArgs | options.DropIndexesArgs |
+		options.DropSearchIndexArgs | options.EncryptArgs |
+		options.EstimatedDocumentCountArgs | options.FindArgs |
+		options.FindOneArgs | options.FindOneAndDeleteArgs |
+		options.FindOneAndReplaceArgs | options.FindOneAndUpdateArgs |
+		options.GridFSFindArgs | options.GridFSNameArgs | options.GridFSUploadArgs |
+		options.IndexArgs | options.InsertManyArgs | options.InsertOneArgs |
+		options.ListCollectionsArgs | options.ListDatabasesArgs |
+		options.ListIndexesArgs | options.ListSearchIndexesArgs |
+		options.LoggerArgs | options.RangeArgs | options.ReplaceArgs |
+		options.RewrapManyDataKeyArgs | options.RunCmdArgs |
+		options.SearchIndexesArgs | options.ServerAPIArgs | options.SessionArgs |
+		options.TimeSeriesArgs | options.TransactionArgs | options.UpdateArgs |
+		options.UpdateSearchIndexArgs | options.ChangeStreamArgs |
+		options.AutoEncryptionArgs
+}
+
+// MongoOptions is an interface that wraps a method to return a list of setter
+// functions for merging options for a generic argument type.
+type MongoOptions[T Args] interface {
+	ArgsSetters() []func(*T) error
+}
+
+// NewArgsFromOptions will functionally merge a slice of mongo.Options in a
+// "last-one-wins" manner, where nil options are ignored.
+func NewArgsFromOptions[T Args](opts ...MongoOptions[T]) (*T, error) {
+	args := new(T)
+	for _, opt := range opts {
+		if opt == nil || reflect.ValueOf(opt).IsNil() {
+			// Do nothing if the option is nil or if opt is nil but implicitly cast as
+			// an Options interface by the NewArgsFromOptions function. The latter
+			// case would look something like this:
+			//
+			// var opt *SomeOptions
+			// NewArgsFromOptions(opt)
+			continue
+		}
+
+		for _, setArgs := range opt.ArgsSetters() {
+			if setArgs == nil {
+				continue
+			}
+
+			if err := setArgs(args); err != nil {
+				return nil, err
+			}
+		}
+	}
+	return args, nil
+}
+
+// Options implements a mongo.Options object for an arbitrary arguments type.
+// The intended use case is to create options from arguments.
+type Options[T Args] struct {
+	Args *T             // Arguments to set on the option type
+	Clbk func(*T) error // A callback for further modification
```

Comment:
Optional: It's not obvious that `Clbk` is an abbreviation for "Callback" because it's similar to other more common abbreviations (e.g. `clk`, `blk`). Consider renaming to `Callback`.